rom/laghos_rom.cpp

```diff
@@ -338,6 +339,84 @@ void ComputeElementRowOfG_E(const IntegrationRule *ir,
     }
 }
 
+// Sets the rows of constraints matrix G for the energy-conserving EQP rule. 
+void ComputeRowsOfG(const IntegrationRule *ir,
+					hydrodynamics::QuadratureData const& quad_data,
+					Vector const& w_e, Vector const& v_e,
+					FiniteElement const& test_fe,
+					FiniteElement const& trial_fe,
+					const int zone_id,
+					const bool equationV, const bool equationE,
+					Vector & rv, Vector & re)
+{
+	const int nqp = ir->GetNPoints();
+	const int dim = trial_fe.GetDim(); // TODO: shouldn't it be the dimension of test_fe?
+	const int h1dofs_cnt = test_fe.GetDof();
+	const int l2dofs_cnt = trial_fe.GetDof();
+
+	if (equationV)
+	{
+		MFEM_VERIFY(rv.Size() == nqp, "");
+	}
+
+	if (equationE)
+	{
+		MFEM_VERIFY(re.Size() == nqp, "");
+		MFEM_VERIFY(v_e.Size() == h1dofs_cnt*dim, "");
+		MFEM_VERIFY(w_e.Size() == l2dofs_cnt, "");
+	}
+
+	DenseMatrix grad_vshape(h1dofs_cnt, dim); // grad of velocity basis vector
+	DenseMatrix loc_force(h1dofs_cnt, dim);
+	Vector Vloc_force(loc_force.Data(), h1dofs_cnt * dim);
+
+	for (int q = 0; q < nqp; q++)
+	{
+		const IntegrationPoint &ip = ir->IntPoint(q);
+
+		// Form stress:grad_vshape at the current point.
+		// TODO: how do we set which velocity basis vector is used for the
+		// gradient calculation?
+		test_fe.CalcDShape(ip, grad_vshape);
+		for (int i = 0; i < h1dofs_cnt; i++)
+		{
+			for (int vd = 0; vd < dim; vd++) // Velocity components.
+			{
+				loc_force(i, vd) = 0.0;
+				for (int gd = 0; gd < dim; gd++) // Gradient components.
+				{
+					loc_force(i, vd) +=
+						quad_data.stressJinvT(vd)(zone_id*nqp + q, gd) * grad_vshape(i,gd);
+				}
+			}
+		}
+
+		// NOTE: UpdateQuadratureData includes ip.weight as a factor in quad_data.stressJinvT,
+		// set by LagrangianHydroOperator::UpdateQuadratureData.
+		loc_force *= 1.0 / ip.weight;  // Divide by exact quadrature weight
+
+		if (equationV)
+		{
+			// set rv = sum(Vloc_force) 
+			rv[q] = 0.0;
+			for (int i = 0; i < h1dofs_cnt * dim; i++)
+			{
+				rv[q] += Vloc_force(i);
+			}
+		}
+		if (equationE)
+		{
+			// set re = F^T * v; F = integral of (stress : grad w_i)phi_j
+			re[q] = 0.0;
+			for (int i = 0; i < l2dofs_cnt; i++)
+			{
+				re[q] += w_e(i);
+			}
+			re[q] *= v_e * Vloc_force;
+		}
```

Comment:
For the energy equation I take the inner product of `Vloc_force` with the velocity snapshot vector `v_e` and multiply it with the sum of `w_e`, with `w_e` an energy basis vector. This is to form the product `F^T * v`, where

- `F` is integral of `(stress : grad w_j) phi_i`, with `w_j` a velocity basis vector and `phi_i` an energy basis vector;
- `v` is a velocity snapshot vector. 

I tried to keep the approach similar to that of the current EQP implementation, but I'm not sure if this yields the correct result.

Reply:
I think what we desire should be `F^T * v = (stress : grad v) phi_i`. There should be no velocity basis in the energy conservation equation.

Reply:
I believe that this is what I'm computing. My understanding is that `v_e * grad w_j` yields the gradient of the velocity snapshot we want, `grad v`. The reason I compute it this way is that it allows me to: 

1. use just one loop to calculate `Vloc_force` for both energy and velocity; 
2. use the given `CalcDShape` function to get the gradient of the basis vector, `grad w_j`. Once I compute `grad w_j`, then I multiply it by the velocity snapshot coefficients, `v_e`, to get the desired `grad v`. Otherwise, I don't know how I can compute the gradient of the snapshot, `grad v`, directly. 

I believe that this is also the way the velocity force vector `F^T * v` is originally constructed in the paper by Dobrev et al (2012), where they outline the underlying FEM discretization.

Reply:
Ah, I see. I overlooked the multiplication. I guess I am following the calculation now, and @dylan-copeland can further confirm it.

Reply:
@dylan-copeland